### DISCUSSION TITLE
Problem using DS18B20 sensor with ESP32 board

### DISCUSSION THREAD

**Discussion by villacampa** [lcgamboa/picsimlab#97] [Q&A] [answered]
First of all, thanks so much for all the great efforts with PICSimLab development.

I'm developing a device full of sensors and software automation based on the ESP32, and tried to replicate my hardware with PICSimLab for testing the firmware on it before deploying it to real hardware.

I got a lot of problems that I'm trying to solve bit by bit. Some of them seems to be solved using the latest unstable release and were related to SDCARD. Thanks once again.

One of the issues I'm not being able to fix and not sure where is the problem, is making a DS18B20 sensor work.
I replicated the most simple example I found in the picsimlab_examples repo, that is working with Arduino One board, but no luck with ESP32.

I uploaded the code I tested here:
https://github.com/villacampa/picsimlab_examples/pull/1

I can confirm the code works on real ESP32 hardware with several DSB18B20 sensors attached to it.

Am I missing any configuration or execution arguments for making it work?

If something is failing in PICSimLab, and I'm given some clues of what to read, check and change, I could try to fix the problem and generate a pull request with the fix.

Thanks!!

**Comment by lcgamboa** [comment, marked as answer]
Hi @villacampa ,

This problem is because Qemu does not support accurate cycle simulation. Therefore, signals that depend on exact timing, such as the onewire, dhtxx, or ws2812b protocols, do not work well with Qemu.  In PICSimLab Using a value for the "Qemu CPU mips" combobox other than "no limit" causes Qemu to try to simulate an execution speed, which causes these peripherals to work sporadically. Setting the option to "Auto" in the example of the DS1820 that you send, some values are received (usually wrong). With the DHT22 it works with the "auto" option and as the IO clock at 4MHz on my computer (example attached). I didn't add examples of these protocols because I still haven't found the best way to deal with the problem.

[NO_ESP32_DHT22.pzw.zip](https://github.com/lcgamboa/picsimlab/files/12798189/NO_ESP32_DHT22.pzw.zip)

**Reply by villacampa** [reply]
Thanks @lcgamboa

Now I understand what's going on, as I'm using other sensors with those protocols and I did not have luck with the QEMU settings.
I will make a bit of research to see if I can help anyhow.

Also, reading the PICSimLab code, I found that in
https://github.com/lcgamboa/picsimlab/blob/master/src/parts/input_ds18b20.h
this code is probably wrong and you would like to name the constant as PART_DS18B20_H instead
```
#ifndef PART_DHT22_H
#define PART_DHT22_H
```

**Reply by lcgamboa** [reply]
Thank you for the warning. I will make the correction. As for the ds1820b, I will check the [timing code ](https://github.com/lcgamboa/picsimlab/blob/a59077cf18286acfddd10299c9f4663680e30538/src/devices/sen_ds18b20.cc#L93)to see if it is possible to make it less susceptible to the timing variations of the Qemu simulation.
